File: src/main/java/br/com/rotaractsorocabaleste/rotasales/common/jwt/JWTAuthenticationFilter.java
package br.com.rotaractsorocabaleste.rotasales.common.jwt;

import br.com.rotaractsorocabaleste.rotasales.user.User;
import br.com.rotaractsorocabaleste.rotasales.user.UserDetailsImpl;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;
import java.util.Map;

import static br.com.rotaractsorocabaleste.rotasales.common.jwt.SecurityConstants.EXPIRATION_TIME;
import static br.com.rotaractsorocabaleste.rotasales.common.jwt.SecurityConstants.SECRET;

@RequiredArgsConstructor
public class JWTAuthenticationFilter extends UsernamePasswordAuthenticationFilter {

    private AuthenticationManager authenticationManager;

    public JWTAuthenticationFilter(final AuthenticationManager authenticationManager) {
        this.authenticationManager = authenticationManager;

        setFilterProcessesUrl("/auth");
    }

    @Override
    public Authentication attemptAuthentication(final HttpServletRequest req,
                                                final HttpServletResponse res) throws AuthenticationException {
        try {
            final var creds = new ObjectMapper()
                    .readValue(req.getInputStream(), User.class);

            return authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(
                            creds.getUsername(),
                            creds.getPassword(),
                            new ArrayList<>())
            );
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected void successfulAuthentication(final HttpServletRequest req,
                                            final HttpServletResponse res,
                                            final FilterChain chain,
                                            final Authentication auth) throws IOException {
        final var objectMapper = new ObjectMapper();
        final var token = JWT.create()
                .withSubject(((UserDetailsImpl) auth.getPrincipal()).getUsername())
                .withExpiresAt(new Date(System.currentTimeMillis() + EXPIRATION_TIME))
                .sign(Algorithm.HMAC512(SECRET.getBytes()));
        final var response = Map.of("access_token", token);

        res.setContentType("application/json");
        res.getWriter().write(objectMapper.writeValueAsString(response));
        res.getWriter().flush();
    }
}

File: src/main/kotlin/br/com/rotaractsorocabaleste/rotasales/entrypoint/config/JWTAuthenticationFilter.kt
package br.com.rotaractsorocabaleste.rotasales.entrypoint.config

import br.com.rotaractsorocabaleste.rotasales.core.entity.User
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter

import javax.servlet.ServletException

import java.io.IOException

import javax.servlet.FilterChain

import javax.servlet.http.HttpServletResponse

import javax.servlet.http.HttpServletRequest

import org.springframework.security.authentication.UsernamePasswordAuthenticationToken

import com.fasterxml.jackson.databind.ObjectMapper
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import java.lang.RuntimeException
import java.util.*


class JWTAuthenticationFilter(
    private val expiration: Long = 60000,
    private val secret: String
) : UsernamePasswordAuthenticationFilter() {

    @Throws(AuthenticationException::class)
    override fun attemptAuthentication(
        req: HttpServletRequest,
        res: HttpServletResponse?
    ): Authentication? {
        return try {
            val credentials: User = ObjectMapper()
                .readValue(req.inputStream, User::class.java)
            authenticationManager.authenticate(
                UsernamePasswordAuthenticationToken(
                    credentials.username,
                    credentials.password,
                    ArrayList()
                )
            )
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    @Throws(IOException::class, ServletException::class)
    override fun successfulAuthentication(
        req: HttpServletRequest?,
        res: HttpServletResponse,
        chain: FilterChain?,
        auth: Authentication
    ) {
        val token: String = Jwts.builder()
            .setSubject((auth.principal as User).username)
            .setExpiration(Date(System.currentTimeMillis() + expiration))
            .signWith(SignatureAlgorithm.HS512, secret.toByteArray())
            .compact()
    }

}